network/src/testutils/test_node.rs

```diff
@@ -0,0 +1,313 @@
+    }
+
+    fn network_ids(&self) -> Vec<NetworkId> {
+        self.peer_network_ids().keys().copied().collect()
+    }
+
+    fn role(&self) -> RoleType {
+        match self.node_id().node_type {
+            NodeType::Validator => RoleType::Validator,
+            _ => RoleType::FullNode,
+        }
+    }
+
+    fn peer_role(&self) -> PeerRole {
+        match self.node_id().node_type {
+            NodeType::Validator => PeerRole::Validator,
+            NodeType::ValidatorFullNode => PeerRole::ValidatorFullNode,
+            NodeType::PublicFullNode => PeerRole::Unknown,
+        }
+    }
+
+    /// Connects a node to another node.  The other's inbound handle must already be added.
+    fn connect(&self, network_id: NetworkId, metadata: ConnectionMetadata) {
+        let self_peer_id = self.peer_network_id(network_id).peer_id();
+        let self_metadata = self.conn_metadata(network_id, ConnectionOrigin::Inbound, None);
+        let remote_peer_id = metadata.remote_peer_id;
+        // Tell the other node it's good to send to the connected peer now
+        self.other_inbound_handle(PeerNetworkId::new(network_id, remote_peer_id))
+            .connect(remote_peer_id, network_id, self_metadata);
+        // Then connect us
+        self.node_inbound_handle(network_id)
+            .connect(self_peer_id, network_id, metadata);
+    }
+
+    /// Disconnects a node from another node
+    fn disconnect(&self, network_id: NetworkId, metadata: ConnectionMetadata) {
+        let self_peer_id = self.peer_network_id(network_id).peer_id();
+        let self_metadata = self.conn_metadata(network_id, ConnectionOrigin::Inbound, None);
+        let remote_peer_id = metadata.remote_peer_id;
+        // Tell the other node it's disconnected
+        self.other_inbound_handle(PeerNetworkId::new(network_id, remote_peer_id))
+            .disconnect(remote_peer_id, network_id, self_metadata);
+        // Then disconnect us
+        self.node_inbound_handle(network_id)
+            .disconnect(self_peer_id, network_id, metadata);
+    }
+
+    /// Find a common network between nodes, they're sorted in priority order
+    fn find_common_network(&self, other: &Self) -> Option<NetworkId> {
+        match self.node_type() {
+            NodeType::Validator => match other.node_type() {
+                NodeType::Validator => Some(NetworkId::Validator),
+                NodeType::ValidatorFullNode => Some(NetworkId::Vfn),
+                NodeType::PublicFullNode => None,
+            },
+            NodeType::ValidatorFullNode => match other.node_type() {
+                NodeType::Validator => Some(NetworkId::Vfn),
+                _ => Some(NetworkId::Public),
+            },
+            NodeType::PublicFullNode => match other.node_type() {
+                NodeType::Validator => None,
+                _ => Some(NetworkId::Public),
+            },
+        }
+    }
+
+    /// Build `ConnectionMetadata` for a connection on another node
+    fn conn_metadata(
```

Comment:
Likewise: do we ever plan on overriding this? If not, let's make it a helper. Also, maybe call it `new_connection_metadata`?